Adding and Updating Functions dealing with Items in Rooms and held by NPCs so that Duplicate Items Cannot Exist within a Singular Hash Table

bvvill opened the pull request:
As described in #1614

bvvill committed:
Making changes stated in #1614

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
merging dev into branch

bvvill committed:
updating move_npc() function description

bvvill committed:
reformatted item_init()

bvvill committed:
same thing as last but better

bvvill committed:
changing `get_item_from_hash()` to `get_item_from_npc()` in the GIVE_ITEM case of `do_node_actions()`

junasano commented on include/npc/npc.h:
```diff
@@ -258,6 +269,26 @@ int add_item_to_npc(npc_t *npc, item_t *item);
+/* Deletes and frees all items in an NPC's inventory
+ * 
+ * Parameters:
+ *  npc: the npc whose inventory we're freeing
```
add deleting as well, so becomes "... inventory we're deleting and freeing"
To make your intentions stated in the issue more explicit.

bvvill replied:
i agree, will do

bvvill committed:
removing redundant / obselete tests regarding npc movement from test_rooms_npc.c

bvvill committed:
Revert "removing redundant / obselete tests regarding npc movement from test_rooms_npc.c", did this on the wrong branch lol

This reverts commit e08c7f5b73fda50c2a8d5d8ce9f5a0dbd6762999.

junasano commented on src/game-state/src/game.c:
```diff
@@ -21,6 +21,22 @@ game_t *game_new(char *desc)
 
     /* read from the file using interface from WDL team */
 
+    room_hash_t *rooms = NULL;
```
Im not sure if it works like this but is there a reason why you stated each variable before hand and set that to NULL?  If not, maybe just go straight into game->all_rooms = NULL; and etc.

bvvill replied:
https://github.com/uchicago-cs/chiventure/issues/1614#issuecomment-1137860606

junasano commented on src/game-state/src/item.c:
```diff
@@ -17,7 +17,14 @@ int item_init(item_t *new_item, char *item_id, char *short_desc,
     strcpy(new_item->long_desc, long_desc);
-    new_item->stat_effects = NULL;
+
+    game_action_hash_t *actions = NULL;
```
Same thing here with declaring the NULL beforehand.

bvvill replied:
https://github.com/uchicago-cs/chiventure/issues/1614#issuecomment-1137860606

junasano commented on tests/npc/test_npc.c:
```diff
@@ -301,16 +300,14 @@ Test(npc, remove_item_from_npc)
     rc = add_item_to_npc(npc, dup_item);
-    cr_assert_eq(rc, SUCCESS, "add_item_to_npc failed to "
-                 "add an item to npc");
+    cr_assert_eq(rc, FAILURE, "add_item_to_npc added duplicate item to npc");
```
I feel like testing for duplicate items on the remove_item_from_npc test is redundant. Since it's already being tested in add_item_to_npc and this test is specifically for removing the item rather than adding.

bvvill replied:
Yeah I agree, removing this as well

junasano commented on tests/game-state/test_rooms.c:
```diff
@@ -120,7 +120,7 @@ Test(room_item, add_duplicate_item_to_room)
 
     int check = add_item_to_room(new_room, test_item2);
-    cr_assert_eq(check, SUCCESS, "duplicate item not added to room correctly");
+    cr_assert_eq(check, FAILURE, "duplicate item added to room incorrectly");
```
isn't this check and the check right below this pretty much doing the same thing?

bvvill replied:
Yup you're right, didn't think about that, the second one shall be removed

junasano commented on tests/game-state/test_rooms.c:
```diff
@@ -168,16 +168,15 @@ Test(room_item, remove_item_from_room)
     rc = add_item_to_room(room, dup_item);
-    cr_assert_eq(rc, SUCCESS, "add_item_to_room failed to "
-                 "add an item to room");
+    cr_assert_eq(rc, FAILURE, "add_item_to_room added duplicate item");
```
Same thing as comment from the other test file. No need to test for duplicate again and since the test is for the remove item test.

bvvill replied:
Yeah I agree, I will remove it now

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

junasano commented on src/game-state/src/room.c:
```diff
@@ -345,63 +355,38 @@ int auto_gen_movement(npc_t *npc, room_list_t *all_rooms)
     }
-    else
+    char *old_room_id = npc->movement->track;
+    assert(move_npc(npc) != 0);
```
change the 0 to SUCCESS so that it is more explicit

bvvill replied:
omg this is actually a major bug lol thank you for pointing this out, since the refactoring of `move_npc_mov()` to simply out `FAILURE` or `SUCCESS`, this statement was not updated to say `assert(move_npc(npc) != FAILURE);`, so I will change that now

junasano commented on include/npc/npc.h:
```diff
@@ -311,6 +342,23 @@ int add_battle_to_npc(npc_t *npc, stat_t *stats, move_t *moves,
+ * technically change rooms.
+ *
+ */
+int move_npc(npc_t *npc);
```
I'm not really sure what the point of this function is except that it is in the npc.h/npc.c module rather than npc_move.h/npc_move.c. Especially since they are pretty much identical and is only used once so far (and in that place, move_npc_mov can be used).

bvvill replied:
Yeah this function is purely for minimizing complexity where possible. I think people are going to more often use functions dealing with higher-level structs such as an `npc_t` (as opposed to its movement struct/member) and a function for moving an NPC given solely it's `npc_t` didn't exist yet, so I figured I'd quickly implement it. It's also a lot easier to read/decipher when used in other functions, and I don't think performance loss is an issue. Also, while right now the function only makes changes to an NPC's movement member, perhaps developers in the future could add more functions inside this one that change something else related to the NPC if they move rooms, such as dialogue or appearance, but this honestly had nothing to do with why I decided to implement `move_npc()`.

junasano commented on src/game-state/src/room.c:
```diff
@@ -17,6 +17,9 @@ int room_init(room_t *new_room, char *room_id, char *short_desc,
     strncpy(new_room->short_desc, short_desc, strlen(short_desc)+1);
     strncpy(new_room->long_desc, long_desc, strlen(long_desc)+1);
 
+    item_hash_t *head = NULL;
```
same thing will NULL here

bvvill replied:
https://github.com/uchicago-cs/chiventure/issues/1614#issuecomment-1137860606

junasano commented on src/npc/src/npc.c:
```diff
@@ -17,6 +17,8 @@ int npc_init(npc_t *npc, char *npc_id, char *short_desc, char *long_desc,
     npc->hostility_level = hostility_level;
     npc->npc_battle = NULL;
     npc->movement = movement;
+    item_hash_t *items = NULL;
```
NULL once again

bvvill replied:
https://github.com/uchicago-cs/chiventure/issues/1614#issuecomment-1137860606

junasano commented on src/npc/src/npc_battle.c:
```diff
@@ -48,9 +48,7 @@ npc_battle_t *npc_battle_new(stat_t* stats, move_t* moves,
-    if (npc_battle == NULL || npc_battle->stats == NULL ||
-            npc_battle->moves == NULL || npc_battle->class_type == NULL ||
-            npc_battle->items == NULL || check != SUCCESS)
+    if (npc_battle == NULL || check != SUCCESS)
```
Is there a reason why you deleted the NULL checks for the other stuff? Because I feel like a battle can't start if those are NULL and just checking to see if npc_battle == NULL isn't enough to check for everything

bvvill replied:
checking npc_battle for being NULL is checking if the memory allocation of the struct worked, and then checking if `check` equals `SUCCESS` makes sure that the initialization of the struct went okay. In some of the examples/tests, there are many parameters that `NULL` is being passed into, and because of that, the return value for `npc_battle_new()` is NULL, which obviously wasn't the intention of whoever was creating those battle structs in those cases. It's also to stay consistent with the other struct creation implementations throughout chiventure, as I haven't come across any other `..._new()` function that expects every single struct member to not be `NULL` and returns `NULL` if not.

junasano commented on src/npc/src/rooms_npc.c:
```diff
@@ -5,7 +5,9 @@ int npcs_in_room_init(npcs_in_room_t *npcs_in_room, char* room_id)
     strncpy(npcs_in_room->room_id, room_id, MAX_ID_LEN);
-    npcs_in_room->npc_list = NULL;
+
+    npc_hash_t *npcs = NULL;
```
NULL here too

bvvill replied:
https://github.com/uchicago-cs/chiventure/issues/1614#issuecomment-1137860606

junasano reviewed (changes requested):
Overall, the logic seems good. I left some comments here and there but I'm not 100% sure for some of them so feel free to correct me if you had a different intention.

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
implementing Jun's suggestions/observations

junasano commented:
The changes look good. Just make sure to merge with the dev branch since there seems to have been new PRs approved.

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

lilyehsani commented on src/game-state/src/room.c:
```diff
@@ -67,11 +70,19 @@ int room_free(room_t *room)
 {
-    int rc;
-
-    rc = add_item_to_hash(&(room->items), item);
```
What is the reason you moved away from using this helper function? I notice that the implementation here is very similar to the implementation for add_item_to_npc, so a helper function makes sense to me.

bvvill replied:
While the helper function is correctly used here, I think that it's worth noting that it assumes the hash handle being used is `hh`, and because there are two possible hash handles for items, I thought it was better to be explicit as to which one was being used. That being said, I think I've become overly paranoid about hash table errors, so if you feel this was an unnecessary change, I will swiftly undo it.

lilyehsani reviewed (changes requested):
Looks really good! One comment.

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

bvvill committed:
changes suggested from lily

bvvill committed:
changed the add_item_to_hash function so that it doesn't add duplicate items

bvvill committed:
lol whoops

bvvill committed:
deleting now unnecessary tests and restoring the functionality of `copy_item_to_hash()` so that it can still add duplicate items

bvvill committed:
whoops 2

bvvill committed:
fixing typo :/

bvvill committed:
adjusting test

bvvill committed:
adjusting test comment

lilyehsani committed:
Merge branch 'dev' into npc/1614-adding-and-updating-functions-dealing-with-items-in-rooms-and-held-by-npcs

lilyehsani reviewed (approved):
LGTM! Approved.

bvvill commented:
comments from #1614:

There are currently a few bugs involving items within rooms or held by NPCs, one being that duplicate items are allowed to be added to a room's or NPC's `item_hash_t`, which is an error according to UTHash's documentation. There are also a couple of useful functionalities that have yet to be implemented, such as "getting" an item from an NPC's inventory given the `item_id`. There are also a couple hash implementations regarding items that could use optimization

As I said I would do in my opening comment, I've implemented `item_t *get_item_from_npc(npc_t *npc, char *item_id)` in npc.h/.c, which returns an item in an NPC's inventory given the `item_id`, if it exists.

I've also implemented `delete_all_items_from_npc` and `free_all_items_from_npc`, which delete and (delete, but also) free the given NPC's inventory, respectively. The difference in the use of these functions is that solely deleting an NPC's inventory doesn't free the items that were in it, so that the items can still be used in other places or conveniently added to other hash tables, as is done in `transfer_all_npc_items()` (found in room.h/.c)

in `game.c` and `item.c`, I've also went ahead and explicitly initialized every hash table member of the `game_t` and `item_t` structs to be `NULL` when either of those two structs is created. Doing so is incredibly important for ut_hash table maintenance and stability, and can lead to mysterious seg faults if not done properly. The new lines added to `game_new()` are:
```
    room_hash_t *rooms = NULL;
    item_hash_t *items = NULL;
    quest_hash_t *quests = NULL;
    npc_hash_t *npcs = NULL;
    stats_global_hash_t *stats = NULL;
    effects_global_hash_t *effects = NULL;
    class_hash_t *classes = NULL;

    game->all_rooms = rooms;
    game->all_items = items;
    game->all_quests = quests;
    game->all_npcs = npcs;
    game->curr_stats = stats;
    game->all_effects = effects;
    game->all_classes = classes;
```
And the new lines in `item_init()` are:
```
    game_action_hash_t *actions = NULL;
    attribute_hash_t *attributes = NULL;
    effects_hash_t *stat_effects = NULL;

    new_item->actions = actions;
    new_item->attributes = attributes;
    new_item->stat_effects = stat_effects;
```


in `add_item_to_npc()` (in npc.h/.c), I've also explicitly implemented the function instead of using `add_item_to_hash()`, since it assumes the hash handle to be hh and doesn't check that the input parameters aren't NULL.

in `game-state/test_rooms.c` and `npc/test_npc.c`, I've updated the tests regarding adding duplicate items to rooms and NPCs, such that the functions now appropriately return `FAILURE` if a duplicate item is attempted to be added

lilyehsani commented:
Issue Score: Satisfactory

Comments:
Great work on this PR! The closing statement is very thorough. Two comments:
1. The summary for this PR is too terse. A good rule of thumb is that you should be writing at least one full paragraph in your summary (see https://uchicago-cs.github.io/cmsc22000/project/issue-pr-grading.html#descriptive-summary for examples).
2. When you create a PR, you are requesting a review of the code in the branch you want to merge. As such, you should not push any commits to the branch right after creating the PR; you should only do so in response to a code review. Make sure that, when you create a PR, the code is ready for review and passes all the tests. Merging dev is acceptable, but there are other commits after the creation of the PR and before the review.

<!-- {"score": "satisfactory", "minor": false} -->
